#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
gRPC server for bazi-rule-service.
"""

from __future__ import annotations
import logging

logger = logging.getLogger(__name__)

import json
import os
import sys
from concurrent import futures

import grpc

BASE_DIR = os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
PROJECT_ROOT = BASE_DIR
sys.path.insert(0, PROJECT_ROOT)

# å¯¼å…¥ç”Ÿæˆçš„ gRPC ä»£ç 
sys.path.insert(0, os.path.join(PROJECT_ROOT, "proto", "generated"))
import bazi_rule_pb2
import bazi_rule_pb2_grpc

from core.calculators.BaziCalculator import BaziCalculator


class BaziRuleServicer(bazi_rule_pb2_grpc.BaziRuleServiceServicer):
    """å®ç° BaziRuleService çš„ gRPC æœåŠ¡"""

    def MatchRules(self, request: bazi_rule_pb2.BaziRuleMatchRequest, context: grpc.ServicerContext) -> bazi_rule_pb2.BaziRuleMatchResponse:
        """åŒ¹é…è§„åˆ™"""
        import datetime
        request_time = datetime.datetime.now().strftime("%Y-%m-%d %H:%M:%S")
        rule_types_str = ", ".join(request.rule_types) if request.rule_types else "å…¨éƒ¨"
        logger.info(f"[{request_time}] ğŸ“¥ bazi-rule-service: æ”¶åˆ°è¯·æ±‚ - solar_date={request.solar_date}, solar_time={request.solar_time}, gender={request.gender}, rule_types=[{rule_types_str}], use_cache={request.use_cache}", flush=True)
        
        try:
            import time
            total_start = time.time()
            
            # 1. å…«å­—è®¡ç®—ï¼ˆä½¿ç”¨æœ¬åœ°è®¡ç®—ï¼Œå’Œæœ¬åœ°åŒ¹é…é€»è¾‘å®Œå…¨ä¸€è‡´ï¼‰
            calc_start = time.time()
            calculator = BaziCalculator(request.solar_date, request.solar_time, request.gender)
            # ç›´æ¥æœ¬åœ°è®¡ç®—ï¼Œä¸è°ƒç”¨å¾®æœåŠ¡ï¼ˆé¿å…å¾ªç¯è°ƒç”¨å’Œæ€§èƒ½é—®é¢˜ï¼‰
            calculator.calculate()
            calc_time = time.time() - calc_start
            logger.info(f"[{request_time}] âœ… bazi-rule-service: å…«å­—è®¡ç®—å®Œæˆï¼ˆè€—æ—¶ {calc_time:.2f}ç§’ï¼‰", flush=True)
            
            # 2. æ„å»ºè§„åˆ™è¾“å…¥ï¼ˆå’Œæœ¬åœ°åŒ¹é…é€»è¾‘å®Œå…¨ä¸€è‡´ï¼‰
            build_start = time.time()
            bazi_data = calculator.build_rule_input()
            build_time = time.time() - build_start
            
            # 3. è§„åˆ™åŒ¹é…ï¼ˆç›´æ¥ä½¿ç”¨æœ¬åœ°åŒ¹é…é€»è¾‘ï¼Œå’Œ _match_rules_locally å®Œå…¨ä¸€æ ·ï¼‰
            # å…³é”®ï¼šä¸è°ƒç”¨ calculator.match_rules()ï¼Œå› ä¸ºå®ƒä¼šå…ˆå°è¯•è°ƒç”¨å¾®æœåŠ¡ï¼Œå¯¼è‡´æ€§èƒ½é—®é¢˜
            # ç›´æ¥è°ƒç”¨ RuleService.match_rules()ï¼Œå’Œæœ¬åœ°åŒ¹é…é€»è¾‘å®Œå…¨ä¸€è‡´
            match_start = time.time()
            rule_types = list(request.rule_types) if request.rule_types else None
            # å¼ºåˆ¶å¯ç”¨ç¼“å­˜ï¼Œé™¤éæ˜ç¡®æŒ‡å®š use_cache=False
            use_cache_optimized = request.use_cache if request.use_cache is False else True
            
            # ç›´æ¥è°ƒç”¨ RuleService.match_rulesï¼Œå’Œæœ¬åœ°åŒ¹é…é€»è¾‘å®Œå…¨ä¸€è‡´
            from server.services.rule_service import RuleService
            matched = RuleService.match_rules(
                bazi_data,
                rule_types=rule_types,
                use_cache=use_cache_optimized
            )
            
            # è·å–æœªåŒ¹é…çš„è§„åˆ™ï¼ˆå’Œæœ¬åœ°åŒ¹é…é€»è¾‘ä¸€è‡´ï¼‰
            engine = RuleService.get_engine()
            all_rules = [r for r in engine.rules if r.get('enabled', True)]
            if rule_types:
                rule_types_set = set(rule_types)
                all_rules = [r for r in all_rules if r.get('rule_type') in rule_types_set]
            matched_rule_ids = {r.get('rule_id') or r.get('rule_code') for r in matched if isinstance(r, dict)}
            unmatched = [r for r in all_rules if (r.get('rule_id') or r.get('rule_code')) not in matched_rule_ids]
            
            match_time = time.time() - match_start
            logger.info(f"[{request_time}] âœ… bazi-rule-service: è§„åˆ™åŒ¹é…å®Œæˆ - åŒ¹é… {len(matched)} æ¡ï¼ŒæœªåŒ¹é… {len(unmatched)} æ¡ï¼ˆè€—æ—¶ {match_time:.2f}ç§’ï¼Œæ„å»º {build_time:.2f}ç§’ï¼Œç¼“å­˜={use_cache_optimized}ï¼‰", flush=True)

            # å°†å¯¹è±¡è½¬æ¢ä¸ºå¯åºåˆ—åŒ–çš„ dictï¼ˆä¼˜åŒ–ï¼šåªåºåˆ—åŒ–å¿…è¦å­—æ®µï¼Œå‡å°‘æ•°æ®é‡ï¼‰
            matched_serializable = []
            for rule in matched:
                if isinstance(rule, dict):
                    # åªä¿ç•™å¿…è¦å­—æ®µï¼Œå‡å°‘åºåˆ—åŒ–æ•°æ®é‡
                    rule_dict = {
                        'rule_id': rule.get('rule_id'),
                        'rule_code': rule.get('rule_code'),
                        'rule_name': rule.get('rule_name'),
                        'rule_type': rule.get('rule_type'),
                        'content': rule.get('content'),
                        'priority': rule.get('priority'),
                    }
                    matched_serializable.append(rule_dict)
                else:
                    matched_serializable.append(dict(rule) if hasattr(rule, '__dict__') else {})
            
            # ä¼˜åŒ–ï¼šä¸è¿”å› unmatched æ•°æ®ï¼Œå‡å°‘å“åº”å¤§å°
            # unmatched æ•°æ®é€šå¸¸å¾ˆå¤§ï¼ˆ436æ¡ï¼‰ï¼Œä½†å®¢æˆ·ç«¯é€šå¸¸åªéœ€è¦ matched æ•°æ®
            unmatched_serializable = []
            # å¦‚æœéœ€è¦ unmatched æ•°æ®ï¼Œå¯ä»¥åç»­é€šè¿‡å•ç‹¬çš„æ¥å£è·å–
            # è¿™é‡Œåªè¿”å›ä¸€ä¸ªç»Ÿè®¡ä¿¡æ¯
            unmatched_count = len(unmatched)
            
            # ä¼˜åŒ–ï¼šå‡å°‘ context æ•°æ®é‡ï¼ˆåªä¿ç•™åŒ¹é…çš„è§„åˆ™ï¼‰
            context_optimized = {}
            if isinstance(calculator.last_rule_context, dict):
                # åªä¿ç•™åŒ¹é…è§„åˆ™çš„ context
                matched_rule_ids = {rule.get('rule_id') or rule.get('rule_code') for rule in matched if isinstance(rule, dict)}
                for rule_id, context_value in calculator.last_rule_context.items():
                    if rule_id in matched_rule_ids:
                        context_optimized[rule_id] = context_value

            response = bazi_rule_pb2.BaziRuleMatchResponse()
            
            # åºåˆ—åŒ–å‰è®°å½•æ—¶é—´
            serialize_start = datetime.datetime.now()
            matched_json_str = json.dumps(matched_serializable, ensure_ascii=False, default=str)
            # unmatched_json_str ä¸å†éœ€è¦ï¼Œå› ä¸ºåªè¿”å› count
            context_json_str = json.dumps(context_optimized, ensure_ascii=False, default=str)
            serialize_end = datetime.datetime.now()
            serialize_time = (serialize_end - serialize_start).total_seconds()
            
            # è®°å½•å“åº”æ•°æ®å¤§å°
            matched_size = len(matched_json_str.encode('utf-8'))
            unmatched_size = 0  # ä¸å†è¿”å› unmatched æ•°æ®
            context_size = len(context_json_str.encode('utf-8'))
            total_size = matched_size + unmatched_size + context_size
            
            response.matched_json = matched_json_str
            # åªè¿”å› unmatched æ•°é‡ï¼Œä¸è¿”å›å®Œæ•´æ•°æ®
            response.unmatched_json = json.dumps({'count': unmatched_count}, ensure_ascii=False)
            response.context_json = context_json_str
            
            metadata = {
                "service": "bazi-rule-service",
                "version": "1.0.0",
            }
            response.metadata_json = json.dumps(metadata, ensure_ascii=False)
            
            total_time = time.time() - total_start
            logger.info(f"[{request_time}] âœ… bazi-rule-service: å“åº”å·²è¿”å›ï¼ˆæ€»è€—æ—¶ {total_time:.2f}ç§’ï¼Œè®¡ç®— {calc_time:.2f}ç§’ï¼ŒåŒ¹é… {match_time:.2f}ç§’ï¼Œåºåˆ—åŒ– {serialize_time:.2f}ç§’ï¼Œå“åº”å¤§å°: matched={matched_size/1024/1024:.2f}MB, unmatched={unmatched_size/1024/1024:.2f}MB, context={context_size/1024/1024:.2f}MB, æ€»è®¡={total_size/1024/1024:.2f}MBï¼‰", flush=True)
            
            # å¼ºåˆ¶åˆ·æ–°è¾“å‡ºï¼Œç¡®ä¿æ—¥å¿—åŠæ—¶å†™å…¥
            import sys
            sys.stdout.flush()
            sys.stderr.flush()
            
            return response
            
        except ValueError as e:
            logger.info(f"[{request_time}] âŒ bazi-rule-service: å‚æ•°é”™è¯¯ - {str(e)}", flush=True)
            context.set_code(grpc.StatusCode.INVALID_ARGUMENT)
            context.set_details(str(e))
            return bazi_rule_pb2.BaziRuleMatchResponse()
        except Exception as e:
            import traceback
            error_msg = f"è§„åˆ™åŒ¹é…å¤±è´¥: {str(e)}\n{traceback.format_exc()}"
            logger.info(f"[{request_time}] âŒ bazi-rule-service: é”™è¯¯ - {error_msg}", flush=True)
            context.set_code(grpc.StatusCode.INTERNAL)
            context.set_details(f"è§„åˆ™åŒ¹é…å¤±è´¥: {str(e)}")
            return bazi_rule_pb2.BaziRuleMatchResponse()

    def HealthCheck(self, request: bazi_rule_pb2.HealthCheckRequest, context: grpc.ServicerContext) -> bazi_rule_pb2.HealthCheckResponse:
        """å¥åº·æ£€æŸ¥"""
        return bazi_rule_pb2.HealthCheckResponse(status="ok")


def serve(port: int = 9004):
    """å¯åŠ¨ gRPC æœåŠ¡å™¨ï¼ˆæ”¯æŒçƒ­æ›´æ–°ï¼‰"""
    try:
        from server.hot_reload.microservice_reloader import (
            create_hot_reload_server,
            register_microservice_reloader
        )
        
        server_options = [
            ('grpc.keepalive_time_ms', 300000),
            ('grpc.keepalive_timeout_ms', 20000),
            ('grpc.keepalive_permit_without_calls', False),
            ('grpc.http2.max_pings_without_data', 2),
            ('grpc.http2.min_time_between_pings_ms', 60000),
            ('grpc.http2.min_ping_interval_without_data_ms', 300000),
            # å¢åŠ æ¶ˆæ¯å¤§å°é™åˆ¶ï¼ˆé»˜è®¤4MBï¼Œå¢åŠ åˆ°50MBä»¥æ”¯æŒå¤§é‡è§„åˆ™ï¼‰
            ('grpc.max_send_message_length', 50 * 1024 * 1024),  # 50MB
            ('grpc.max_receive_message_length', 50 * 1024 * 1024),  # 50MB
        ]
        
        server, reloader = create_hot_reload_server(
            service_name="bazi_rule",
            module_path="services.bazi_rule.grpc_server",
            servicer_class_name="BaziRuleServicer",
            add_servicer_to_server_func=bazi_rule_pb2_grpc.add_BaziRuleServiceServicer_to_server,
            port=port,
            server_options=server_options,
            max_workers=20,
            check_interval=30
        )
        
        register_microservice_reloader("bazi_rule", reloader)
        reloader.start()
        
        # create_hot_reload_server å·²ç»ç»‘å®šäº†ç«¯å£ï¼Œä¸éœ€è¦å†æ¬¡ç»‘å®š
        server.start()
        logger.info(f"âœ… Bazi Rule gRPC æœåŠ¡å·²å¯åŠ¨ï¼ˆçƒ­æ›´æ–°å·²å¯ç”¨ï¼‰ï¼Œç›‘å¬ç«¯å£: {port}")
        
        try:
            server.wait_for_termination()
        except KeyboardInterrupt:
            logger.info("\n>>> æ­£åœ¨åœæ­¢æœåŠ¡...")
            reloader.stop()
            server.stop(grace=5)
            logger.info("âœ… æœåŠ¡å·²åœæ­¢")
            
    except ImportError:
        # é™çº§åˆ°ä¼ ç»Ÿæ¨¡å¼
        server_options = [
            ('grpc.keepalive_time_ms', 300000),
            ('grpc.keepalive_timeout_ms', 20000),
            ('grpc.keepalive_permit_without_calls', False),
            ('grpc.http2.max_pings_without_data', 2),
            ('grpc.http2.min_time_between_pings_ms', 60000),
            ('grpc.http2.min_ping_interval_without_data_ms', 300000),
            ('grpc.max_send_message_length', 50 * 1024 * 1024),
            ('grpc.max_receive_message_length', 50 * 1024 * 1024),
        ]
        
        server = grpc.server(
            futures.ThreadPoolExecutor(max_workers=20),
            options=server_options
        )
        bazi_rule_pb2_grpc.add_BaziRuleServiceServicer_to_server(BaziRuleServicer(), server)
        
        listen_addr = f"[::]:{port}"
        server.add_insecure_port(listen_addr)
        
        server.start()
        logger.info(f"âœ… Bazi Rule gRPC æœåŠ¡å·²å¯åŠ¨ï¼ˆä¼ ç»Ÿæ¨¡å¼ï¼‰ï¼Œç›‘å¬ç«¯å£: {port}")
        
        try:
            server.wait_for_termination()
        except KeyboardInterrupt:
            logger.info("\n>>> æ­£åœ¨åœæ­¢æœåŠ¡...")
            server.stop(grace=5)
            logger.info("âœ… æœåŠ¡å·²åœæ­¢")


if __name__ == "__main__":
    import argparse
    parser = argparse.ArgumentParser(description="å¯åŠ¨ Bazi Rule gRPC æœåŠ¡")
    parser.add_argument("--port", type=int, default=9004, help="æœåŠ¡ç«¯å£ï¼ˆé»˜è®¤: 9004ï¼‰")
    args = parser.parse_args()
    serve(args.port)

